gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/NumberHelper.java

```diff
@@ -721,6 +723,86 @@ public static Number coerceTo(final Number a, final Class<? extends Number> claz
         return a;
     }
 
+    /**
+     * Casts the given number to the specified numeric type if it can fit into it.
+     * Otherwise, throw.
+     *
+     * @param a the number to be cast
+     * @param numberToken the number token denoting the desired type to cast
+     * @return the number cast to the specified type
+     * @throws IllegalArgumentException if the specified numeric type is unsupported
+     * @throws ArithmeticException if the number overflows
+     */
+    public static Number castTo(final Number a, final N numberToken) {
+        Class<?> clazz = numberToken.getType();
+        if (a.getClass().equals(clazz)) {
+            return a;
+        } else if (clazz.equals(Integer.class)) {
+            Long val = getLong(a, numberToken);
+            if (val >= Integer.MIN_VALUE && val <= Integer.MAX_VALUE) {
+                return a.intValue();
+            }
+        } else if (clazz.equals(Long.class)) {
+            return getLong(a, numberToken);
+        } else if (clazz.equals(Float.class)) {
+            // BigDecimal to double will overflow into Infinity, we want to throw instead of passing through
+            if (!a.getClass().equals(BigDecimal.class) &&
+                    (Double.isInfinite(a.doubleValue()) || Double.isNaN(a.doubleValue())))  {
+                return a.floatValue();
+            }
+            if (a.doubleValue() >= -Float.MAX_VALUE && a.doubleValue() <= Float.MAX_VALUE) {
+                return a.floatValue();
+            }
+        } else if (clazz.equals(Double.class)) {
+            // BigDecimal to double will overflow into Infinity,  we want to throw instead of passing through
+            if (!a.getClass().equals(BigDecimal.class) &&
+                    (Double.isInfinite(a.doubleValue()) || Double.isNaN(a.doubleValue()))) {
+                return a.doubleValue();
+            }
+            if (!Double.isInfinite(a.doubleValue())) {
+                // float losses precision, use string intermediate
+                return a.getClass().equals(Float.class) ? Double.parseDouble(a.toString()) : a.doubleValue();
+            }
+        } else if (clazz.equals(Byte.class)) {
+            Long val = getLong(a, numberToken);
+            if (val >= Byte.MIN_VALUE && val <= Byte.MAX_VALUE) {
+                return a.byteValue();
+            }
+        } else if (clazz.equals(Short.class)) {
+            Long val = getLong(a, numberToken);
+            if (val >= Short.MIN_VALUE && val <= Short.MAX_VALUE) {
+                return a.shortValue();
+            }
+        } else if (clazz.equals(BigInteger.class)) {
+            return NumberHelper.bigIntegerValue(a);
```

Comment:
There appears to be a preexisting bug in how `NumberHelper.bigIntegerValue(a)` handles large floats or doubles:
```
gremlin> g.inject(new Double("1000000000000000000000")).asNumber(N.nbigInt)
==>9223372036854775807
```
If the double or float is beyond the range of long, the conversion to BigInteger results in a value of long max or long min.